From patches Mon Sep 17 00:00:00 2001
From: fred.konrad@greensocs.com
To: Anthony Liguori <aliguori@us.ibm.com>,
 Peter Maydell <peter.maydell@linaro.org>,
 =?UTF-8?B?S09OUkFEIEZyw6lkw6lyaWM=?= <fred.konrad@greensocs.com>,
 qemu-devel@nongnu.org, fred.konrad@greensocs.com
Cc: Anthony Liguori <aliguori@us.ibm.com>, peter.maydell@linaro.org,
 mark.burton@greensocs.com, qemu-devel@nongnu.org,
 Luiz Capitulino <lcapitulino@redhat.com>, cornelia.huck@de.ibm.com,
 afaerber@suse.de, fred.konrad@greensocs.com
Date: Fri, 12 Apr 2013 00:38:59 +0200
Message-id: 1365719939-7169-1-git-send-email-fred.konrad@greensocs.com
Subject: [PATCH 1/1] virtio-balloon: fix dynamic properties.

From: KONRAD Frederic <fred.konrad@greensocs.com>

To keep compatibility with the old virtio-balloon-x, add the dynamic properties
to virtio-balloon-pci and virtio-balloon-ccw.

Cc: Luiz Capitulino <lcapitulino@redhat.com>
Signed-off-by: KONRAD Frederic <fred.konrad@greensocs.com>
---
 hw/s390x/virtio-ccw.c              |  8 ++++++++
 hw/virtio/virtio-balloon.c         | 16 ++++++++--------
 hw/virtio/virtio-pci.c             |  8 ++++++++
 include/hw/virtio/virtio-balloon.h |  9 +++++++++
 4 files changed, 33 insertions(+), 8 deletions(-)

diff --git a/hw/s390x/virtio-ccw.c b/hw/s390x/virtio-ccw.c
index 4dec0cd..ea5ed6d 100644
--- a/hw/s390x/virtio-ccw.c
+++ b/hw/s390x/virtio-ccw.c
@@ -626,6 +626,14 @@ static void virtio_ccw_balloon_instance_init(Object *obj)
     VirtIOBalloonCcw *dev = VIRTIO_BALLOON_CCW(obj);
     object_initialize(OBJECT(&dev->vdev), TYPE_VIRTIO_BALLOON);
     object_property_add_child(obj, "virtio-backend", OBJECT(&dev->vdev), NULL);
+
+    object_property_add(obj, "guest-stats", "guest statistics",
+                        balloon_stats_get_all, NULL, NULL, &dev->vdev, NULL);
+
+    object_property_add(obj, "guest-stats-polling-interval", "int",
+                        balloon_stats_get_poll_interval,
+                        balloon_stats_set_poll_interval,
+                        NULL, &dev->vdev, NULL);
 }
 
 static int virtio_ccw_scsi_init(VirtioCcwDevice *ccw_dev)
diff --git a/hw/virtio/virtio-balloon.c b/hw/virtio/virtio-balloon.c
index c2c446e..2e2ea30 100644
--- a/hw/virtio/virtio-balloon.c
+++ b/hw/virtio/virtio-balloon.c
@@ -105,8 +105,8 @@ static void balloon_stats_poll_cb(void *opaque)
     virtio_notify(vdev, s->svq);
 }
 
-static void balloon_stats_get_all(Object *obj, struct Visitor *v,
-                                  void *opaque, const char *name, Error **errp)
+void balloon_stats_get_all(Object *obj, struct Visitor *v, void *opaque,
+                           const char *name, Error **errp)
 {
     VirtIOBalloon *s = opaque;
     int i;
@@ -129,17 +129,17 @@ static void balloon_stats_get_all(Object *obj, struct Visitor *v,
     visit_end_struct(v, errp);
 }
 
-static void balloon_stats_get_poll_interval(Object *obj, struct Visitor *v,
-                                            void *opaque, const char *name,
-                                            Error **errp)
+void balloon_stats_get_poll_interval(Object *obj, struct Visitor *v,
+                                     void *opaque, const char *name,
+                                     Error **errp)
 {
     VirtIOBalloon *s = opaque;
     visit_type_int(v, &s->stats_poll_interval, name, errp);
 }
 
-static void balloon_stats_set_poll_interval(Object *obj, struct Visitor *v,
-                                            void *opaque, const char *name,
-                                            Error **errp)
+void balloon_stats_set_poll_interval(Object *obj, struct Visitor *v,
+                                     void *opaque, const char *name,
+                                     Error **errp)
 {
     VirtIOBalloon *s = opaque;
     int64_t value;
diff --git a/hw/virtio/virtio-pci.c b/hw/virtio/virtio-pci.c
index 2b22588..9a3d291 100644
--- a/hw/virtio/virtio-pci.c
+++ b/hw/virtio/virtio-pci.c
@@ -1450,6 +1450,14 @@ static void virtio_balloon_pci_instance_init(Object *obj)
     VirtIOBalloonPCI *dev = VIRTIO_BALLOON_PCI(obj);
     object_initialize(OBJECT(&dev->vdev), TYPE_VIRTIO_BALLOON);
     object_property_add_child(obj, "virtio-backend", OBJECT(&dev->vdev), NULL);
+
+    object_property_add(obj, "guest-stats", "guest statistics",
+                        balloon_stats_get_all, NULL, NULL, &dev->vdev, NULL);
+
+    object_property_add(obj, "guest-stats-polling-interval", "int",
+                        balloon_stats_get_poll_interval,
+                        balloon_stats_set_poll_interval,
+                        NULL, &dev->vdev, NULL);
 }
 
 static const TypeInfo virtio_balloon_pci_info = {
diff --git a/include/hw/virtio/virtio-balloon.h b/include/hw/virtio/virtio-balloon.h
index 3b459bb..023d0f0 100644
--- a/include/hw/virtio/virtio-balloon.h
+++ b/include/hw/virtio/virtio-balloon.h
@@ -69,4 +69,13 @@ typedef struct VirtIOBalloon {
     int64_t stats_poll_interval;
 } VirtIOBalloon;
 
+void balloon_stats_get_all(Object *obj, struct Visitor *v, void *opaque,
+                           const char *name, Error **errp);
+void balloon_stats_get_poll_interval(Object *obj, struct Visitor *v,
+                                     void *opaque, const char *name,
+                                     Error **errp);
+void balloon_stats_set_poll_interval(Object *obj, struct Visitor *v,
+                                     void *opaque, const char *name,
+                                     Error **errp);
+
 #endif
-- 
1.8.1.4

